lib/src/grpc_api_impl/grpc.dart

```diff
@@ -36,12 +36,12 @@ class Channel extends protobuf.RpcClient {
 
   Channel(this.servicePackage, this.client);
 
-  Future<protobuf.GeneratedMessage> invoke(
+  Future<T> invoke<T extends protobuf.GeneratedMessage>(
```

Comment:
Why do we need this? It makes things just slower. The caller can do the down-cast on the call site.

Reply:
This solution was preferred internally, in cl/196552547. We have an external pending PR for what you're suggesting (I think): dart-lang/dart-protoc-plugin#104.

I had a brief chat with @lrhn about this recently, and IIRC he favored this solution as well, rather than unwrapping, casting, and rewrapping the return value, but if it's slower, we should change it.

We seem to be using generic methods everywhere these days. Are they inherently slower, or is it "just" a matter of optimizing them?

Reply:
We should rely on data more than on my thinking, so we should just measure.

Generally speaking, the VM has to perform checks on incoming argument values against the pararameter declared types.

The overhead of generic methods is as follows:
   * The caller must pass an additional type arguments vector
   * The callee has to do a `arg is T` check on function entry

Now with this code here, we have the overhead 2 times, once for calling `Channel.invoke<T>()` and another time for `Client.invoke<T>()`.

The alternative would be as follows:

```dart
class GeneratedApiClient {
  Future<FooReponse> foo(FooRequest fooRequest) async {
    final FooResponse fooResponse = new FooResponse();
    await _channel.invoke('service', 'foo-method', fooRequest, fooResponse);
    return response;
  }
}
```

Notice that we don't even have to cast any return value, because the return value of `_channel.invoke()` is the same as `fooResponse` (why does it even need to return the `response`, it could just be `Future<Null>`)

Since there are no checks necessary in this case, it seems to me this is strictly less work.

Reply:
I ran a very unscientific benchmark, comparing just the overhead from `invoke()`. The castless version is indeed faster than the genericized version:
```
$ dart --preview-dart-2 benchmarks/rpc_client.dart 
Genericized(RunTime): 3.401904386075325 us.
Castless(RunTime): 3.3208579768669035 us.
```

Since the difference isn't greater, and the genericized version was already committed internally and synced to package:protobuf, I'm going to land this as is, and open an issue over in dart-lang/protobuf.